File: MySQLBackup.pl
#!/usr/bin/perl

#############################################################################
# mysqlbackup.pl  version .08-beta                                          #
#                                                                           #
# History and information:                                                  #
# http://www.ghostwheel.com/merlin/Personal/notes/mysqlbackup/              #
#                                                                           #
# Synapsis:                                                                 #
#   This script allows you to make date-time stamped backups of individual  #
#   mysql databases.  Each database backup is a gzip compressed sql script  #
#   that can be restored/imported via the mysql command line tool.          #
#                                                                           #
#   The tables of each database being backed up are locked, preserving data #
#   integrity.  While this does temporarily prevent writes to the tables,   #
#   the impact is reduced by backing up each database separately.           #
#                                                                           #
#   This script backups all tables accessible to the mysql user that it is  #
#   configured to use.  That means this script is useful to the sysadmin as #
#   well as to regular users who whish to maintain backups of their own     #
#   mysql databases.                                                        #
#                                                                           #
#   This script requires that your mysql connection information be stored   #
#   in a my.cnf or .my.cnf file in the home directory of the user used to   #
#   run this script.   Because it makes an external call to mysqldump, it   #
#   is inadvisible to hard code the username and password into this script  #
#   and then call mysqldump with user and pass as command line arguments.   #
#   Doing that would expose your mysql username and password to anyone      #
#   logged into the server while the backup was occuring.  For information  #
#   on how to create a my.cnf file visit this page:                         #
#   http://dev.mysql.com/doc/refman/5.0/en/option-files.html                #
#                                                                           #
# Usage:                                                                    #
#   mysqlbackup.pl BACKUPDIR                                                #
#                                                                           #
#                                                                           #
#############################################################################
#                                                                           #
#  History:                                                                 #
#                                                                           #
#  .08-beta  Utilizes [client] section of ~/.my.cnf if present.             #
#                                                                           #
#  .07-beta  Fixed bad logic in the utility file path code.                 #
#            Added a set of common path locations to the search path.       #
#            Removed information_schema from databases backed up.           #
#                                                                           #
#  .06-beta  Modified sql connect string to no longer specify an initial    #
#            database.  Checks to make sure there is a [mysqldump]          #
#            section in the ~/.my.conf file.                                #
#                                                                           #
#  .05-beta  Initial public release.  First release without hard coded      #
#            values for utility locations, usernames, passwords and hosts.  #
#                                                                           #
#                                                                           #
#############################################################################
##
##
## * Copyright (c) 2008,2009, Chris O'Halloran
## * All rights reserved.
## *
## * Redistribution and use in source and binary forms, with or without
## * modification, are permitted provided that the following conditions are met:
## *     * Redistributions of source code must retain the above copyright
## *       notice, this list of conditions and the following disclaimer.
## *     * Redistributions in binary form must reproduce the above copyright
## *       notice, this list of conditions and the following disclaimer in the
## *       documentation and/or other materials provided with the distribution.
## *     * Neither the name of Chris O'Halloran nor the
## *       names of any contributors may be used to endorse or promote products
## *       derived from this software without specific prior written permission.
## *
## * THIS SOFTWARE IS PROVIDED BY Chris O'Halloran ''AS IS'' AND ANY
## * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
## * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
## * DISCLAIMED. IN NO EVENT SHALL Chris O'Halloran BE LIABLE FOR ANY
## * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
## * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
## * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
## * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
## * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
## * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

use DBI;
use POSIX qw(strftime);

## Here is an example of how to specify a location for a external utilities used by this script.  
#$UtilLocation{'gzip'} = '/usr/bin/gzip';
#$UtilLocation{'mysqldump'} = '/usr/local/bin/mysqldump';


## Let's make sure there is a good PATH in place when this script runs:
$ENV{'PATH'} .= ':/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/opt/bin:/opt/sbin:/opt/local/bin:/opt/local/sbin:~/bin';

## Locate the following utilities for use by the script
@Utils = ('gzip', 'mysqldump');
foreach $Util (@Utils) 
	{

    ##  Populate $UtilLocation{$Util} if it isn't set manually
    if ( !(defined($UtilLocation{$Util})) ) 
		{
		($UtilLocation{$Util} = `which $Util`) =~ s/[\n\r]*//g;
		}      

	## If $UtilLocation{$Util} is still not set, we have to abort.	
	if ( !(defined($UtilLocation{$Util})) || $UtilLocation{$Util} eq "" )
		{
		die("Unable to find $Util in the current PATH.\n");
		}
	elsif ( !(-f $UtilLocation{$Util}) )
		{
		die("$UtilLocation{$Util} is not valid.\n");
		}

	}



$BackupTimeStamp = strftime("%Y-%m-%d_%H-%M", localtime);

if ( defined($ARGV[0]) ) {
	($BackupDir = $ARGV[0]) =~ s/\/$//g;
	if ( !(-d $BackupDir) ) {
		die("$BackupDir is not a directory.\n");
	}
	elsif ( !(-w $BackupDir) ) {
		die("$BackupDir is not not writable.\n");
	}
}
else {
	die("USAGE:  MySQLBackup.pl DESTDIR\n");
}

my %MySQLInfo = &ParseMyCnf;
if ('' eq $MySQLInfo{'filename'} ) {
	die("Unable to locate my.cnf file.\nFor information on how to create a my.cnf file visit this page:\nhttp://dev.mysql.com/doc/refman/5.0/en/option-files.html\n\n");
}

my $DBUser;
my $DBPass;
my $DBHost = 'localhost';
my $DBPort = '3306';

if ('' ne $MySQLInfo{'mysqldump'}{'user'} &&  '' ne $MySQLInfo{'mysqldump'}{'pass'}) {
	$DBUser = $MySQLInfo{'mysqldump'}{'user'};
	$DBPass = $MySQLInfo{'mysqldump'}{'pass'};
	if ('' ne $MySQLInfo{'mysqldump'}{'host'}) {
		$DBHost = $MySQLInfo{'mysqldump'}{'host'};
	}
	if ('' ne $MySQLInfo{'mysqldump'}{'port'}) {
		$DBPort = $MySQLInfo{'mysqldump'}{'port'};
	}
}
elsif ('' ne $MySQLInfo{'client'}{'user'} &&  '' ne $MySQLInfo{'client'}{'password'}) {
	$DBUser = $MySQLInfo{'client'}{'user'};
	$DBPass = $MySQLInfo{'client'}{'password'};
	if ('' ne $MySQLInfo{'client'}{'host'}) {
		$DBHost = $MySQLInfo{'client'}{'host'};
	}
	if ('' ne $MySQLInfo{'client'}{'port'}) {
		$DBPort = $MySQLInfo{'client'}{'port'};
	}
}
else {
	die("Unable to determine login credentials for mysqldump.\nYou must have a [mysqldump] or [client] section with values for user and pass.\nFor information on how to create a my.cnf file visit this page:\nhttp://dev.mysql.com/doc/refman/5.0/en/option-files.html\n\n");
}




my $dbh1 = DBI->connect("DBI:mysql::$DBHost:$DBPort", $DBUser, $DBPass, { AutoCommit => 1 }) || die($DBI::errstr);
   
   
my $ListSearch = $dbh1->prepare("show databases;") || &DBIConnectFailure($DBI::errstr);

$ListSearch->execute;
while ( ($temp1) = $ListSearch->fetchrow_array)
  {
  if ( $temp1 ne "information_schema" )
  	{
	  print "Backing Up $temp1\n";
	  $Backup = "$BackupDir/$temp1-$BackupTimeStamp.sql.gz";
	  system("$UtilLocation{'mysqldump'} --defaults-extra-file=$MySQLInfo{'filename'} --hex-blob --opt --lock-tables $temp1 | $UtilLocation{'gzip'} -c > $Backup");
	 }
  }
$ListSearch->finish;
   
exit;




sub ParseMyCnf {
##  ParseMyCnf
##
##  This is a simple subroutine for parsing the contents of a my.cnf file so that an enlightened MySQL user
##  or administrator doesn't have to hard code their credentials into one of my administrative utils.
##

	
	my %MyConfData;
	my $group;
	my $opt_name;
	my $value;
	my $MyCnfTmp;
	
	my $HomeDir = glob("~");
	foreach $MyCnfTmp ( ("~/my.cnf", "~/.my.cnf" ) ) {
		my $MyCnfTest;
		($MyCnfTest = $MyCnfTmp) =~ s/~/$HomeDir/ex;
		if ( -f $MyCnfTest  && '' eq $MyConfData{'filename'} ) {
			$MyConfData{'filename'} = $MyCnfTest;
		}
	}

	if ( '' ne $MyConfData{'filename'} ) {
		open(MYCNF, $MyConfData{'filename'}) ||  die("Unable to open $MyConfData{'filename'} file\n");
		while (<MYCNF>) {
			my $Line;
			($Line = $_) =~ s/[\n\r]+//g;
		
			if ($Line =~ m/\[([^\]]+)\]/) {
				$group = $1;
			}

			if ($Line =~ m/^(.*)=(.*)$/) {
				$opt_name = $1;
				$value = $2;
				if ( $value =~ m/^['"](.+)['"]$/ ) {
					$value = $1;
				}
				$MyConfData{$group}{$opt_name} = $value;
			}
		}
		close(MYCNF);
	}
	
	return(%MyConfData);
}

File: MySQLBackup.py
#!/usr/bin/env python3

##############################################################################
# MySQLBackup.py  version 1.0                                                #
#                                                                            #
# Python port of MySQLBackup.pl                                              #
#                                                                            #
# Usage:                                                                     #
#   python MySQLBackup.py DESTDIR                                            #
#   python3 MySQLBackup.py DESTDIR                                           #
#                                                                            #
# Requires no external Python packages.  Uses the mysql and mysqldump       #
# command-line clients, which must already be installed on the system.      #
#                                                                            #
# Your MySQL credentials must be stored in ~/my.cnf or ~/.my.cnf with a    #
# [mysqldump] section (user + password) or a [client] section (user + password).#
# http://dev.mysql.com/doc/refman/5.0/en/option-files.html                  #
##############################################################################

import sys
import os
import subprocess
import shutil
import configparser
from datetime import datetime
from pathlib import Path


def find_utility(name):
    path = shutil.which(name)
    if not path:
        sys.exit("Unable to find {} in the current PATH.".format(name))
    if not os.path.isfile(path):
        sys.exit("{} is not valid.".format(path))
    return path


def parse_my_cnf():
    """Parse ~/my.cnf or ~/.my.cnf, returning a dict keyed by section name."""
    home = Path.home()
    cnf_path = None
    for candidate in [home / 'my.cnf', home / '.my.cnf']:
        if candidate.is_file():
            cnf_path = candidate
            break

    if cnf_path is None:
        return {}

    config = configparser.RawConfigParser(allow_no_value=True)
    try:
        config.read(str(cnf_path))
    except configparser.Error as exc:
        sys.exit("Unable to parse {}: {}".format(cnf_path, exc))

    result = {'filename': str(cnf_path)}
    for section in config.sections():
        result[section] = dict(config.items(section))
    return result


def main():
    # Augment PATH with common utility locations (mirrors Perl script)
    extra_paths = [
        '/sbin', '/bin', '/usr/sbin', '/usr/bin', '/usr/games',
        '/usr/local/sbin', '/usr/local/bin',
        '/opt/bin', '/opt/sbin', '/opt/local/bin', '/opt/local/sbin',
        os.path.expanduser('~/bin'),
    ]
    os.environ['PATH'] = os.environ.get('PATH', '') + ':' + ':'.join(extra_paths)

    gzip_path      = find_utility('gzip')
    mysqldump_path = find_utility('mysqldump')
    mysql_path     = find_utility('mysql')

    if len(sys.argv) < 2:
        sys.exit("USAGE: MySQLBackup.py DESTDIR")

    backup_dir = sys.argv[1].rstrip('/')
    if not os.path.isdir(backup_dir):
        sys.exit("{} is not a directory.".format(backup_dir))
    if not os.access(backup_dir, os.W_OK):
        sys.exit("{} is not writable.".format(backup_dir))

    mysql_info = parse_my_cnf()
    if not mysql_info.get('filename'):
        sys.exit(
            "Unable to locate my.cnf file.\n"
            "For information on how to create a my.cnf file visit:\n"
            "http://dev.mysql.com/doc/refman/5.0/en/option-files.html\n"
        )

    timestamp = datetime.now().strftime('%Y-%m-%d_%H-%M')

    list_result = subprocess.run(
        [mysql_path, '--defaults-extra-file={}'.format(mysql_info['filename']),
         '-N', '-e', 'SHOW DATABASES'],
        stdout=subprocess.PIPE, stderr=subprocess.PIPE,
    )
    if list_result.returncode != 0:
        sys.exit("Failed to list databases: {}".format(list_result.stderr.decode().strip()))
    databases = [db for db in list_result.stdout.decode().splitlines() if db]

    for db_name in databases:
        if db_name == 'information_schema':
            continue

        print("Backing Up {}".format(db_name))
        backup_file = os.path.join(backup_dir, "{}-{}.sql.gz".format(db_name, timestamp))

        dump_cmd = [
            mysqldump_path,
            '--defaults-extra-file={}'.format(mysql_info['filename']),
            '--hex-blob',
            '--opt',
            '--lock-tables',
            db_name,
        ]

        with open(backup_file, 'wb') as out_fh:
            dump_proc = subprocess.Popen(dump_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
            gzip_proc = subprocess.Popen([gzip_path, '-c'], stdin=dump_proc.stdout, stdout=out_fh)
            # Close our copy of dump's stdout so gzip receives EOF when dump exits
            dump_proc.stdout.close()
            gzip_proc.wait()
            dump_stderr = dump_proc.stderr.read()
            dump_proc.wait()

        if dump_proc.returncode != 0:
            print(
                "Warning: mysqldump failed for {}: {}".format(db_name, dump_stderr.decode().strip()),
                file=sys.stderr,
            )


if __name__ == '__main__':
    main()
